### ISSUE TITLE
Trivial Splits in the trees?

### ISSUE THREAD

**Issue by fryasdf** [dmlc/xgboost#1112]
I have happily trained an xgboost ensemble and it performs reasonable well on the test set. However, after looking at the tree in text form I realized that there are multiple lines of the following form (R notation):

```
 [1] "booster[0]"                                   
 [2] "0:[f70<-1.00136e-005] yes=1,no=2,missing=1"   
 [3] "1:[f71<-1.00136e-005] yes=3,no=4,missing=3"   
 [....]
```

feature 70 (which is actually column 71 in the supplied matrix as xgboost starts counting at 0 while R starts counting at 1) is a 0-1-valued feature in my case (coming from the encoding of a boolean feature as two new 0-1-valued dimensions like value_of_feature_is_true and value_of_feature_is_false).

So let me get that right: The top node in this tree asks whether or not this feature has a value smaller than -1.00136*exp(-5). Is that right? Well, this is certainly a trivial split as all the data rows have values 0 or 1 (in particular, positive!)...

The same thing happens with the feature odor=none in the example in the docs:

```
data(agaricus.train, package='xgboost')

# Both dataset are list with two items, a sparse matrix and labels
# (labels = outcome column which will be learned).
# Each column of the sparse Matrix is a feature in one hot encoding format.
train <- agaricus.train

bst <- xgboost(data = train$data, label = train$label, max.depth = 2,
               eta = 1, nthread = 2, nround = 2,objective = "binary:logistic")

# train$data@Dimnames[[2]] represents the column names of the sparse matrix.
xgb.importance(train$data@Dimnames[[2]], model = bst)

# Same thing with co-occurence computation this time
xgb.importance(train$data@Dimnames[[2]], model = bst, data = train$data, label = train$label)

xgb.dump(bst)
```

Here we also get trivial splits. However, since both model perform reasonably well, I guess that these splits make sense and xgboost does secretly apply some transformation to the data (subtract mean or something like this?)?

Is that true? I was not able to find anything useful in the docs concerning this... This value '-1.00136e-005' seems to be special and independent of the actual data set as it occurs quite often in completely different data sets...

What is going on here?

Regards,

fryasdf

**Comment by khotilov** [comment]
This is a model trained on sparse data where xgboost considers sparse "zeros" as "missing" values, and branches them accordingly.

**Comment by fryasdf** [comment]
@khotilov Huh? What do you mean by 'branches them accordingly'? The feature has only two values, 0 and 1. Do you mean that it changes the zeroes in the sense that it puts all zeroes to be some other value?

**Comment by khotilov** [comment]
The values of features in this sparse data from xgboost point of view are "missing" and 1.
E.g., `"0:[f70<-1.00136e-005] yes=1,no=2,missing=1"` shows that the missings go into the yes branch.

It's unfortunate that the split value started being negative for such data. That broke a few things with implicit assumptions that sparsity means zeros in sparse data.

**Comment by fryasdf** [comment]
Aaah, you mean that this split weirdly encodes 'f70 == 0 --> goto 1' and 'f70==1 --> goto 2' because if f70 is 1 then it is not missing but violates the condition (hence goes to 'no', i.e. node 2) and if it is zero, it is being treated as 'NA' meaning it goes to the NA path (i.e. node 1).... I see... Thank you :-)
